### PR TITLE
[Rename] Collection of improvements to make syntactic rename easier to read

### PR DESCRIPTION
**Pull request by ahoppen**
A collection of four independent refactoring that simplify syntactic rename. Should be reviewable commit by commit.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

File: lib/Refactoring/SyntacticRenameRangeDetails.cpp [hunk 1]
```diff
@@ -366,89 +364,151 @@ RegionType RenameRangeDetailCollector::getSyntacticRenameRegionType(
...
+      // happen to have the same name. Don't return them as matching ranges.
+      return RegionType::Unmatched;
+    } else if (specialBaseName != SpecialBaseName::None &&
+               resolved.labelType == LabelRangeType::None) {
+      // Filter out non-semantic special basename locations with no labels.
+      // We've already filtered out those in active code, so these are
+      // any appearance of just 'init', 'subscript', or 'callAsFunction' in
+      // strings, comments, and inactive code.
+      return RegionType::Unmatched;
+    }
```
Comment (hamishknight) [inline lib/Refactoring/SyntacticRenameRangeDetails.cpp:424]: Personally I would write these as consecutive `if` statements given the first one returns, but I don't feel too strongly about it

File: tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp [hunk 2]
```diff
@@ -2583,29 +2586,30 @@ void SwiftLangSupport::findRelatedIdentifiersInFile(
         std::vector<ResolvedLoc> ResolvedLocs = resolveRenameLocations(
             Locs.getLocations(), /*NewName=*/StringRef(), *SrcFile, Diags);
 
-        for (auto ResolvedLoc : ResolvedLocs) {
+        assert(ResolvedLocs.size() == Locs.getLocations().size());
+        for (size_t index = 0; index < ResolvedLocs.size(); ++index) {
+          auto RenameLoc = Locs.getLocations()[index];
```
Comment (hamishknight) [inline tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:2591]: `RenameLoc` doesn't appear to be used for anything here, should it be removed?

**Review by hamishknight — APPROVED** [review]

File: lib/Refactoring/SyntacticRenameRangeDetails.cpp (hunk 1 above)
Comment (ahoppen) [reply lib/Refactoring/SyntacticRenameRangeDetails.cpp:424]: Updated

File: tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp (hunk 2 above)
Comment (ahoppen) [reply tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:2591]: Oh, good catch. Also means that we no longer need `size_t index`.

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

File: lib/Refactoring/SyntacticRenameRangeDetails.cpp [hunk 3]
```diff
@@ -366,89 +364,153 @@ RegionType RenameRangeDetailCollector::getSyntacticRenameRegionType(
...
-                      Resolved.labelType == LabelRangeType::CallArg;
-
-    if (renameLabels(Resolved.labelRanges, Resolved.firstTrailingLabel,
-                     Resolved.labelType, isCallSite))
-      return Config.Usage == NameUsage::Unknown ? RegionType::Unmatched
-                                                : RegionType::Mismatch;
+  if (!handleLabels) {
+    return regionKind;
+  }
+  if (renameLabels(resolved.labelRanges, resolved.firstTrailingLabel,
```
Comment (bnbarham) [inline lib/Refactoring/SyntacticRenameRangeDetails.cpp:507]:
```suggestion
  if (handleLabels && renameLabels(resolved.labelRanges, resolved.firstTrailingLabel,
```

File: tools/SourceKit/tools/sourcekitd/lib/Service/Requests.cpp [hunk 4]
```diff
@@ -3995,25 +3996,16 @@ editorFindModuleGroups(StringRef ModuleName, ArrayRef<const char *> Args) {
...
-                             std::vector<RenameLocations> &RenameLocations,
+                             std::vector<RenameLoc> &RenameLocations,
                              llvm::SmallString<64> &Error) {
   bool Failed = Req.dictionaryArrayApply(KeyRenameLocations,
                                          [&](RequestDict RenameLocation) {
-    int64_t IsFunctionLike = false;
-    if (RenameLocation.getInt64(KeyIsFunctionLike, IsFunctionLike, false)) {
-      Error = "missing key.is_function_like";
-      return true;
-    }
```
Comment (bnbarham) [inline tools/SourceKit/tools/sourcekitd/lib/Service/Requests.cpp:4006]: Have you checked that we send the `()` in all cases that `is_function_like` was set?

File: include/swift/Refactoring/RenameLoc.h [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+/// should find the argument labels as well as some semantic information needed
+/// to resolve the rename ranges.
+struct RenameLoc {
+  /// The line at which the base name starts (1-based).
+  unsigned Line;
+
+  /// The column at which the base name (excluding trivia) starts (1-based).
+  unsigned Column;
+
+  ///  /// The offset at which the related name starts.
```
Comment (bnbarham) [inline include/swift/Refactoring/RenameLoc.h:56]:
```suggestion
  /// The offset at which the related name starts.
```

**Review by bnbarham — APPROVED** [review]

File: tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp [hunk 6]
```diff
@@ -2583,29 +2586,28 @@ void SwiftLangSupport::findRelatedIdentifiersInFile(
...
 
+        assert(ResolvedLocs.size() == Locs.getLocations().size());
         for (auto ResolvedLoc : ResolvedLocs) {
           if (ResolvedLoc.range.isInvalid()) {
             continue;
           }
           unsigned Offset = SrcMgr.getLocOffsetInBuffer(
               ResolvedLoc.range.getStart(), BufferID);
-          Ranges.emplace_back(Offset, ResolvedLoc.range.getByteLength());
+          Ranges.push_back({Offset, ResolvedLoc.range.getByteLength()});
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp:2596]: Nah, can’t be bothered.

File: lib/Refactoring/SyntacticRenameRangeDetails.cpp (hunk 3 above)
Comment (ahoppen) [reply lib/Refactoring/SyntacticRenameRangeDetails.cpp:507]: I prefer my version. `renameLabels` is already such a long call that I don’t want to make the condition even longer, and thus less readable IMO.

File: tools/SourceKit/tools/sourcekitd/lib/Service/Requests.cpp (hunk 4 above)
Comment (ahoppen) [reply tools/SourceKit/tools/sourcekitd/lib/Service/Requests.cpp:4006]: AFAICT we always send `()` when renaming functions. Xcode definitely sends the parentheses.

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by bnbarham** [comment]
@swift-ci please smoke test macOS platform

**Pull request merged**